http-clients/apache-client/src/main/java/software/amazon/awssdk/http/apache/async/internal/InputStreamToPublisherAdapter.java

```diff
@@ -0,0 +1,152 @@
+/*
+ * Copyright 2010-2018 Amazon.com, Inc. or its affiliates. All Rights Reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License").
+ * You may not use this file except in compliance with the License.
+ * A copy of the License is located at
+ *
+ *  http://aws.amazon.com/apache2.0
+ *
+ * or in the "license" file accompanying this file. This file is distributed
+ * on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either
+ * express or implied. See the License for the specific language governing
+ * permissions and limitations under the License.
+ */
+
+package software.amazon.awssdk.http.apache.async.internal;
+
+import java.io.IOException;
+import java.io.InputStream;
+import java.nio.ByteBuffer;
+import java.util.concurrent.BlockingQueue;
+import java.util.concurrent.ExecutorService;
+import java.util.concurrent.Executors;
+import java.util.concurrent.LinkedBlockingQueue;
+import java.util.concurrent.atomic.AtomicReference;
+
+import org.reactivestreams.Publisher;
+import org.reactivestreams.Subscriber;
+import org.reactivestreams.Subscription;
+
+import software.amazon.awssdk.annotations.SdkInternalApi;
+
+@SdkInternalApi
+public class InputStreamToPublisherAdapter {
```

Comment:
Is this our own implementation or is it derived from Rx?

Reply:
Our own but I borrowed ideas around the "signals" and using a thread/queue to serialize the events for this and the `PublishertoInputStreamAdapter` from this class in https://github.com/reactive-streams/reactive-streams-jvm: https://github.com/reactive-streams/reactive-streams-jvm/blob/582fdae4d347d7e07815fb79e2171dd51c37d383/examples/src/main/java/org/reactivestreams/example/unicast/AsyncIterablePublisher.java#L31